idaes/power_generation/carbon_capture/mea_solvent_system/property_package/liquid_prop.py

```diff
@@ -0,0 +1,1282 @@
+
+    def release_state(blk, flags, outlvl=0):
+        '''
+        Method to release state variables fixed during initialisation.
+        Keyword Arguments:
+            flags : dict containing information of which state variables
+                    were fixed during initialization, and should now be
+                    unfixed. This dict is returned by initialize if
+                    hold_state=True.
+            outlvl : sets output level of of logging
+        '''
+
+        if flags is None:
+            return
+
+        # Unfix state variables
+        revert_state_vars(blk, flags)
+
+        init_log = idaeslog.getInitLogger(blk.name, outlvl, tag="properties")
+        init_log.info_high('States released.')
+
+
+@declare_process_block_class("LiquidStateBlock",
+                             block_class=LiquidStateBlockMethods)
+class LiquidStateBlockData(StateBlockData):
+    """
+    Liquid phase property package of amine-based scrubbing of CO2
+    """
+
+    def build(self):
+        """
+        Callable method for Block construction
+        """
+        super(LiquidStateBlockData, self).build()
+
+        # Object reference for molecular weight if needed by CV1D
+        # Molecular weights
+        add_object_reference(self, "mw",
+                             self.config.parameters.mw)
+
+        self._make_state_vars()
+        self._make_speciation_model()
+
+    def _make_speciation_model(self):
+        self.conc_comp_true = Var(self._params.component_list_true,
+                               domain=NonNegativeReals,
+                               initialize=10,
+                               doc="concentration of true species "
+                               "[mol/m3]")
+        def rule_logC(blk,i):
+          return log(blk.conc_comp_true[i])
+
+        self.logC = Expression(self._params.component_list_true,
+                               rule = rule_logC,
+                               doc='logarithm of true species concentration')
+
+        def rule_speciation_model(blk):
+          xa  = blk.conc_comp
+          xt  = blk.conc_comp_true
+          CT  = blk.logC
+          lnk = blk.lnk_eq_conc
+
+          #chemical equilibruim reaction 1
+          yield lnk['R1'] == \
+               CT['MEA+'] + CT['MEACOO-'] - 2*CT['MEA'] - CT['CO2']
+          #chemical equilibruim reaction 2
+          yield lnk['R2'] == \
+               CT['MEA+'] + CT['HCO3-'] - CT['MEA'] -  CT['CO2'] - CT['H2O']
+          #MEA balance
+          yield xa['MEA'] ==  xt['MEA'] + xt['MEA+'] +  xt['MEACOO-']
+          #CO2 balance
+          yield xa['CO2'] ==  xt['CO2'] + xt['HCO3-'] + xt['MEACOO-']
+          #H2O balance
+          yield xa['H2O'] ==  xt['H2O'] + xt['HCO3-']
+          #charge balance
+          yield xt['MEA+'] == xt['HCO3-'] + xt['MEACOO-']
+
+        try:
+            # Try to build speciation model constraints
+            self.speciation_model = ConstraintList(rule=rule_speciation_model)
```

Comment:
it can be written as indexed constraints(in this case index 1 to 6). However, the ContrainList provides a nice way of grouping the equations. Are there concerns with the ContraintList?

Reply:
No concerns, just wondering why as it most of the time we would just create an IndexedConstraint. The only potential concern I can see is knowing how to access each of the `Constraints` in the `ConstraintList` - i.e. what are the individual names.

Reply:
okay, I have used indexed constraints instead.